src/uart.rs

```diff
@@ -0,0 +1,375 @@
+    /// * `main` The mainloop that is executed inside the wrapper
+    ///
+    /// # Examples
+    /// ```
+    /// use riot_wrappers::uart::UartDevice;
+    /// let mut cb = |new_data| {
+    ///     //do something here with the received data
+    /// };
+    /// let mut scoped_main = |self_: &mut UartDevice| loop {
+    ///    self_.write(b"Hello from UART")
+    /// };
+    /// let mut uart = UartDevice::new_scoped(0, 115200, &mut cb, scoped_main)
+    ///    .unwrap_or_else(|e| panic!("Error initializing UART: {e:?}"));
+    /// ```
+    pub fn new_scoped<'scope, F, Main, RMain>(
+        index: usize,
+        baud: u32,
+        user_callback: &'scope mut F,
+        main: Main,
+    ) -> Result<RMain, UartDeviceError>
+    where
+        F: FnMut(u8) + Send + 'scope,
+        Main: FnOnce(&mut Self) -> RMain,
+    {
+        // This possibly relies on Rust code in RIOT to not unwind.
+        let mut self_ = unsafe { Self::construct_uart(index, baud, user_callback) }?;
+        let result = (main)(&mut self_);
+        drop(self_);
+        Ok(result)
+    }
+
+    /// Tries to initialize the given `UART`. Returns a Result with rather `Ok<Self>` if the UART was initialized successfully or a
+    /// `Err<UartDeviceStatus>` containing the error. As the name implies, the created `UART` device can <b>ONLY</b> send data
+    ///
+    /// # Arguments
+    ///
+    /// * `dev` - The index of the hardware device
+    /// * `baud` - The used baud rate
+    ///
+    /// # Examples
+    /// ```
+    /// use riot_wrappers::uart::UartDevice;
+    /// let mut uart = UartDevice::new_without_rx(0, 115200)
+    /// .unwrap_or_else(|e| panic!("Error initializing UART: {e:?}"));
+    /// uart.write(b"Hello from UART");
+    /// ```
+    pub fn new_without_rx(index: usize, baud: u32) -> Result<Self, UartDeviceError> {
+        unsafe {
+            let dev = macro_UART_DEV(index as c_uint);
+            uart_init(dev, baud, None, ptr::null_mut())
+                .negative_to_error()
+                .map(|_| Self { dev })
+                .map_err(UartDeviceError::from_c)
+        }
+    }
+
+    /// Tries to initialize the given `UART` with a static callback. Returns a Result with rather `Ok<Self>` if the UART
+    /// was initialized successfully or a `Err<UartDeviceStatus>` containing the error
+    ///
+    /// # Arguments
+    ///
+    /// * `dev` - The index of the hardware device
+    /// * `baud` - The used baud rate
+    /// * `user_callback` The user defined callback that gets called from the os whenever new data is received from the `UART`
+    ///
+    /// # Examples
+    /// ```
+    /// use riot_wrappers::uart::UartDevice;
+    /// static mut CB: fn(u8) = |new_data| {
+    ///     //do something here with the received data
+    /// };
+    /// let mut uart = UartDevice::new_with_static_cb(0, 115200, unsafe { &mut CB })
```

Comment:
The better example would be to use `static_cell` or maybe `crate::mutex::Mutex`, or even `println!("Received {:02x}", new_data);`, but unless you feel super motivate to update this, we can leave that as is for now.

Reply:
I have added the `println!` in the examples :)

Reply:
Then could almost just be `fn cb(new_data: &[u8) { println!(...) }`, no need for a mut static that'll need unsafe and makes the linter scream.

Almost, because we don't get a `&mut cb`, which hints at that user_callback should be F and not &'scope mut F … which again we can't do because we don't have a size limit.

We may need to iterate over that later on, but that'll work better when we see actual use cases. So let's have this in for now. I've just updated the example to not use static muts any more.